### PR TITLE
Code optimizations reported by static code analysis (2020-05-28) 

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

SUMMARY: Performance "Code optimizations reported by static code analysis"

#### Purpose of change

Cleanup code while increase performance and fix minor errors.

#### Describe the solution

See commit messages.

### CODE REVIEW

File: src/advanced_inv_pane.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ std::vector<advanced_inv_listitem> avatar::get_AIM_inventory( const advanced_inv
         if( worn_item.contents.empty() ) {
             continue;
         }
-        for( std::list<item *> it_stack : item_list_to_stack( worn_item.contents.all_items_top() ) ) {
+        for( const auto &it_stack : item_list_to_stack( worn_item.contents.all_items_top() ) ) {
```
Comment (kevingranade) [inline src/advanced_inv_pane.cpp:122]: Const good, ref good, plsno auto.
```suggestion
        for( const std::list<item *> &it_stack : item_list_to_stack( worn_item.contents.all_items_top() ) ) {
```

File: src/color.cpp [hunk 2]
```diff
@@ -654,7 +654,7 @@ std::string colorize( const translation &text, const nc_color &color )
 
 std::string get_note_string_from_color( const nc_color &color )
 {
-    for( auto i : color_by_string_map ) {
+    for( const auto &i : color_by_string_map ) {
```
Comment (kevingranade) [inline src/color.cpp:657]: Eschew auto
```suggestion
    for( const std::pair<std::string, note_color> &i : color_by_string_map ) {
```

File: src/input.cpp [hunk 3]
```diff
@@ -139,25 +139,25 @@ void input_manager::init()
     }
     t_actions &actions = action_contexts["DEFAULTMODE"];
     std::set<action_id> touched;
-    for( std::map<char, action_id>::const_iterator a = keymap.begin(); a != keymap.end(); ++a ) {
-        const std::string action_id = action_ident( a->second );
+    for( const auto &a : keymap ) {
```
Comment (kevingranade) [inline src/input.cpp:142]:
```suggestion
    for( const std::pair<char, action_id> &a : keymap ) {
```

File: src/color.cpp (hunk 2 above)
Comment (anothersimulacrum) [reply src/color.cpp:657]: Should probably be
```suggestion
    for( const std::pair<const std::string, note_color> &i : color_by_string_map ) {
```
to make clang 10 happy.

File: src/consumption.cpp [hunk 4]
```diff
@@ -550,7 +550,7 @@ int Character::vitamin_mod( const vitamin_id &vit, int qty, bool capped )
 
 void Character::vitamins_mod( const std::map<vitamin_id, int> &vitamins, bool capped )
 {
-    for( auto vit : vitamins ) {
+    for( const auto &vit : vitamins ) {
```
Comment (anothersimulacrum) [inline src/consumption.cpp:553]:
```suggestion
    for( const std::pair<const vitamin_id, int> &vit : vitamins ) {
```

File: src/input.cpp (hunk 3 above)
Comment (anothersimulacrum) [reply src/input.cpp:142]: Again, clang 10 probably wants
```suggestion
    for( const std::pair<const char, action_id> &a : keymap ) {
```

File: src/magic.cpp [hunk 5]
```diff
@@ -1472,7 +1472,7 @@ void known_magic::update_mana( const Character &guy, float turns )
 std::vector<spell_id> known_magic::spells() const
 {
     std::vector<spell_id> spell_ids;
-    for( auto pair : spellbook ) {
+    for( const auto &pair : spellbook ) {
```
Comment (anothersimulacrum) [inline src/magic.cpp:1475]:
```suggestion
    for( const std::pair<const spell_id, spell> &pair : spellbook ) {
```

File: src/npc.h [hunk 6]
```diff
@@ -168,7 +168,7 @@ class job_data
             const std::pair<activity_id, int> &b ) {
                 return a.second > b.second;
             } );
-            for( std::pair<activity_id, int> elem : pairs ) {
+            for( const std::pair<activity_id, int> &elem : pairs ) {
```
Comment (anothersimulacrum) [inline src/npc.h:171]: I suspect clang10 will complain about this
```suggestion
            for( const std::pair<const activity_id, int> &elem : pairs ) {
```

**Comment by ZhilkinSerg** [comment]
I wanted to process `const auto`'s later, but okay.

File: src/npc.h [hunk 7]
```diff
@@ -168,7 +168,7 @@ class job_data
             const std::pair<activity_id, int> &b ) {
                 return a.second > b.second;
             } );
-            for( std::pair<activity_id, int> elem : pairs ) {
+            for( const std::pair<const activity_id, int> &elem : pairs ) {
```
Comment (ZhilkinSerg) [inline src/npc.h:171]:
```suggestion
            for( std::pair<const activity_id, int> &elem : pairs ) {
```

File: src/json.cpp [hunk 8]
```diff
@@ -1241,7 +1241,6 @@ bool JsonIn::get_bool()
     }
     err << "not a boolean value!  expected 't' or 'f' but got '" << ch << "'";
     error( err.str(), -1 );
```
Comment (kevingranade) [inline src/json.cpp:1243]: IIRC one of the static checker will yell at you about this.
```suggestion
    error( err.str(), -1 );
    throw JsonError( "warnings are silly" );
```

Comment (kevingranade) [reply src/json.cpp:1243]: Nevermind, I think I was thinking of somewhere else.

**Pull request merged**